Draft idea for post deserialization validation

woolie opened the pull request:
The remaining issue is how to provide validation on the built in types and whether that is important.
Right now it is easy to add an override for the type you are implementing deserialization for, but I'm not sure how you would do the build ins like Int, Double, String without adding the validation on the type that has those inside of them.

I only added tests the BasicItem and nothing for [BasicItem] nor attributes, but the support code is there and can be added if people like the idea.

woolie commented:
Not sure why I am getting nesting listing errors with the GitHub Action but I can't repro locally. I am using the latest SwiftLint.

drmohundro commented:
@woolie code looks great, thanks! I'll see if I can figure out what is going on with the linter - it might just be the GitHub Actions version.

drmohundro commented:
I'm hoping that if the `.github/workflows/lint.yml` file is updated with the following (just a bump to `3.2.1`) it will resolve the lint issues:

```yml
name: SwiftLint

on:
  [push, pull_request]

jobs:
  docker-lint:
    name: Docker Lint
    runs-on: ubuntu-latest

    steps:
      - uses: actions/checkout@v1
      - name: GitHub Action for SwiftLint with --strict
        uses: norio-nomura/action-swiftlint@3.2.1
        with:
          args: --strict
```

woolie commented:
@drmohundro so I up'ed the lint version as you suggested, I also updated the swift version to current and updated the simulator to the current as well. All of those changes made things pass.

drmohundro commented:
Sorry for taking a few weeks to get it merged... looks great to me. Next steps to get it fully released are to get docs and a version bump. I'm _thinking_ about including #246 in this, too, and then doing a full version bump. That PR is about a rename so that the primary class doesn't match the module name... some weird SPM issue I think.
